### ISSUE TITLE
I suppose that write action doen't work with CNTKTextFormatReader

### ISSUE THREAD

**Issue by Henry-White** [Microsoft/CNTK#495]
Hi,
Latest master. Rebuild.
At the first time I've got strange result on my own data and all I got in output_file just a copy of "labels" from test_file. I'd supposed that may be I do something wrong and I decided to make an experiment on the existing examples.

I used new Simple.cntk (with CNTKTextFormatReader).
Logs with a training and running data are included: [https://cloud.mail.ru/public/7RKD/4FrCYQMTF](url) 
Sorry. Github does not accepted ZIP file for a some reasons. :(

After first run, the first line in SimpleOutput ("labels") contains: 1, 
so that's a result of evaluation for corresponding line from SimpleDataTest_cntk_text.txt
I changed the first line to "|labels 1 0	|features 1 1" which has to be a result equals zero in SimpleOutput, but it does not (after run it remains equal 1). According to this I suppose that the evaluation of NN did not happened.
And I have another confusion about this example. In the config we have:
            labels = [
                dim = 2        # two-dimensional labels
But output contains only one-dimensional label. How to get the second one?  

And I still have a feeling that I do something wrong. If I wrong, could you give some notes about the right thinking direction? Thanks a lot.

And... Thank you very much for this excellent project. The quality of the code is very impressive.

**Comment by raaaar** [comment]
`|labels 1 0	|features -1 -1` -- is the first line in the original SimpleDataTest_cntk_text.txt, which produces `0 // labels` in the output file. If I change it to `|labels 0 1	|features -1 -1`, I get  `1 // labels` as the first line of the output.

**Comment by Henry-White** [comment]
Thanks raaaar,
You are right. Probably I've made mistake with the data-files... Sorry.
But the problem is still remains, not with the SimpleDataTest but my own data (some astronomical data).
I've took SimpleDataTest configuration as a basis and made small changes:
- remove mapping (I have float 100 input and 1 output normalized data)
- change dimensions for layerSizes = 100:30:1 (with one hidden layer - just for simplified calculations)
- change dimensions for reader: features...100, labels...1 and format "dense"
- and change "file" to my own files.

Training without errors.
I had researched a training log and recognized that training process did not pass correctly (CrossEntropyWithSoftmax = 0.00000000 * 250; EvalErrorPrediction = 0.00000000 * 250; till the end).

On evolution I'm using the same method - changing "labels" in the first line. For example, in a TEST-DATA I made the copy of the first line with a small change (input vector is the same):
|Out 0.0008496187746940 and |Out 1.0008496187746940
after evaluation I have in output:
0.000850 and 1.000850.

I.e. just the copy of the "out" of the test-data. May be it's just a result of bad hands.drv :)
Right now I'm trying to resolve this issue through the "NDL" description, but still studding...

**Comment by Henry-White** [comment]
I've made the same test with NDL description. The issue the same:
1) It seems that net does not training at all (CrossEntropyWithSoftmax = 0.00000000 * 250; EvalErrorPrediction = 0.00000000 * 250; - I've got this from the beginning till the end). I don't understand why this happened. I have simple data ( In| 0.1144662022599077 0.1194128174602169 0.1240020590554560 0.1281779693106627... |Out -0.0980508933144290 ). All data has normal distribution close to Gaussian. Dimensions described correctly (100:1).

2) The "write" action just give me a copy of "Test-data" (write any digits in "Test-data" in Out and have the same in output log).

**Comment by frankseide** [comment]
For 1), is that for CPU or GPU? If CPU, please say

    set ACML_FMA=0

and try again. You may be hitting a bug in the ACML library w.r.t. new FMA instructions.

**Comment by Henry-White** [comment]
Hello Frank,
Thank you for spending your time for my problem. I very appreciated for this (I'm studying the code but there are the lots of the code that not so easy to understand for me so far). I checked everything and answer to your question is positive (CPU is used and variable is set). Just to confirm this:
C:\Users\admin\Source\Repos\cntk>set | find "ACML"
ACML_FMA=0
ACML_PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\acml-6.1.0.33-ifort64\ifort64_mp

Unfortunately the issue is the same. And there is some strange behaviour. Because of unchangeable CE and Err during training I could suppose that learning rate is too low and according to the log this would be true:
Finished Epoch[ 1 of 30]: [Training] CE = 0.00000000 * 43309; ErrPredict = 0.00000000 * 43309; totalSamplesSeen = 43309; learningRatePerSample = 0.0089999996; epochTime=1.03429s

but in the config file I pointed out
learningRatesPerMB = 0.9*10:0.5*10:0.3

I've looked through the log file and realised that the all values of LR was reduced to one hundred times.
This is true for Simple Network Builder and for NDL. But because of NDL became deprecated (
I just found out about it) I will remake my sample to BS-model and put result here.

Thank you very much.

**Comment by frankseide** [comment]
The learning rate is actually OK. Notice that it logs learningRatePerSample, while you configure learningRatesPerMB. Please check out [this section](https://github.com/Microsoft/CNTK/wiki/Tutorial2#sgd-parameters) on how learning rates are configured and converted.

**Comment by frankseide** [comment]
But I think this is not your problem. You are starting from a randomly initialized model, so if the model does not update, you would not see an error rate of 0%.

Could you confirm that you are running in CPU (deviceId=-1)? You can grep for PU

    % grep PU G2P_train.log
    Created model with 789 nodes on GPU 0.

If you have a message saying you run on GPU, then the problem may be different. I just discovered that our Windows build does not compile the GPU code for older CUDA cards. It only builds for compute capability of 3.0 and above. Older cards such as a Quadro 6000 have compute capability 2.1, and there seems to be a bug in the NVidia driver in that it accepts the binary (they do some cross-JITting), but then does not execute it correctly. You can verify that by running the Debug build, which for some reason still includes 2.0. If Debug works, and Release does not, then the best solution would be to find a newer graphics card... :(

**Comment by Henry-White** [comment]
I had read this but obviously missed this point "the value defined here will be multiplied by the minibatch size". It is so unusual for me (I have experience with FF, RPROP nets) but right now I think I've understood it. It has been made for MT-purposes. Am I right? It's very simple and very smart decision. I had tried to resolve MT-issue in my own code (RPROP net) but got spooky result.

About my problem... Because of my experience with CUDA I know that this is a big issue with "double"-precision so the first step I did - set up deviceId=-1. 
Right now I use precision = "float" for speed up calculation.

I am ashamed to use your time for my problem. Let I will look BS model. I'll try it. At first sight, it has a more intuitive syntax. If I will get a strange result I will write here. 
And thanks a lot again.

**Comment by frankseide** [comment]
The reason we print it this way is that once we go to sequence-to-sequence models, the number of samples is no longer uniform throughout the network, since inputs and outputs have different lengths. It is also possible that training and eval criteria use different numbers of samples. For example, imagine a multi-task learner that interpolates multiple labels (e.g. multiple languages) where each task has a different number of samples in the minibatch while we want to keep track of each task's performance separately. The output format was inspired by HTK (Hidden Markov Toolkit) which is a popular speech-recognition toolkit (but not relevant in other ML disciplines).

Would you mind trying to use the GPU with precision="float" and let me know which GPU you have? Another thing we can try is to use MKL instead of the ACML library. We are preparing the switch to MKL due to these problems. The current master branch can already build with MKL. I hope we will soon have a downloadable binary with MKL.

You are not taking my time at all. We have seen this problem with several users, and found that there are two bugs, one in ACML with FMA instructions, and one in the NVidia JITter to compute capability 2.1, which both manifest themselves similarly. It is very important to figure this out.

BrainScript is the right thing, but it will not likely solve this problem. Both BrainScript and the old NDL compile into the same internal network representations. The problem you observe is related to execution, particularly an incorrect network definition would not manifest itself in a prediction error of 0.

Thanks for your patience with us!

**Comment by Henry-White** [comment]
Frank, thanks a lot for giving me a clarity. It was very useful.

I have installed MKL (disabled ACML), pulled latest sources and right now I'm rebuilding binaries (MS VS 2013 SP5). About CUDA... I have ability to run CNTK on two computers. One of them have capability 2.1 and another one 3.0. Could you tell me what particular tests do you need and I will do. As far, as I understand, I should to do these tests:
1) deviceId=-1 to see if this situation will be resolved on MKL libs;
2) deviceId=0 and run on the 2.1's device to see if it will affect the result on it.
3) Do you need to check it on 3.0 ?
4) Correct me if I was wrong or just give me know about your needs.

So, the rebuilding process has been finished a minute ago. Has been failed a couple of modules:  LMSequenceReader and CompositeDataReader. It's known issue, connected with version of OpenCV. In project is using 3.0 (opencv_world300.lib) but I have installed 3.1. It's easy to fix but to save time I'll fix it late.

Because NDL out of law I'll use Simple Network Builder but one tricky question still bothering me - the right syntax for "write" action... What should I use as parameter for outputNodeNames, if I do not use mapping? I had used "PosteriorProb : labels", "outputs:labels" and "ScaledLogLikelihood", but this is mostly a guess, not knowledge.  I feel very nervous if i do something I do not understand :) It seems to me that all I'm doing, it just copy the input stream into output. Please, show me correct way.
Thank you very much.

ps. About patience... Actually it is all vice versa. It's thanks to you. You've done and you do a great job which is hardly possible to overestimated for all scientists and researchers around the world.

**Comment by frankseide** [comment]
If you can, please switch to the compute capability 3.0. 2.1 is not supported by our Windows build, and because some functions require NVidia libraries that require 3.0, we have removed support for 2.x just now, coming online soon.

`LMSequenceReader` should have no dependence on OpenCV. Please just try building a second time.

>What should I use as parameter for outputNodeNames

You should list the names of the nodes, surrounded by parentheses, separated by colon (no space). The "write" action will evaluate all these nodes and write their values to different files. To know the names, please look at the training log file, lines like `Validating -->`. This will give you the precise names of all nodes in the system. I believe there should be one named `PosteriorProb`, and that's the one you want.

**Comment by Henry-White** [comment]
I've made execution on 3.0 and made output for other nodes (outputNodeNames = PosteriorProb : EvalErrorPrediction : CrossEntropyWithSoftmax : HLast) just to understand what is going on there.
The result is very strange, for me at least.
All values are the same throughout the output files so for brevity, only write the values themselves:
PosteriorProb = 1.000000 (11775 lines according to the samples in TEST-file)
HLast = -1026536.500000 (11775 lines)
EvalErrorPrediction and CrossEntropyWithSoftmax = 0.0 (just 7 lines and I don't understand why).

I don't see any difference between CPU and GPU (only Hlast changing).

I have tried to execute BS-model but got an error ([https://github.com/Microsoft/CNTK/issues/607](url))

**Comment by Henry-White** [comment]
The most reasons were in using non-symmetric functions, so the issue was in educational level.
If you have the same issue and data distribution is (1, -1 or so), probably you have to use at least TanH for activation function and SquareError for criterion.
Thank for everything. Closed.
